### ISSUE TITLE
Custom pk generation

### ISSUE THREAD

**Issue by jahlfeld** [javers/javers#1277]
**Is your feature request related to a problem? Please describe.**
The current way to generate primary keys does not really work for me as I am using a system with multiple databases and application instances and Sequence is not stateless.

**Describe the solution you'd like**
I tried to make the pk-generation exchangable: If a KeyGenerator is provided when building the JaversSqlRepository, this will be used. Otherwise the previous solution depending on the dialect will be used as a fallback. Thus, nothing will change for applications that already used JaVers before.
Further effects are that it is possible to use the same pk-generation that is used in the rest of the application and we don't necessarily have to use the db to generate our pks.

**Additional context**
PR is on the way. Please let me know your thoughts on this.

**Comment by bartoszwalacik** [comment]
Did you try `RandomCommitGenerator` ?

**Comment by bartoszwalacik** [comment]
Please clarify what feature you actually propose.

**Comment by bartoszwalacik** [comment]
I looked at the code , and I don't see why multiple Javers instances can't use a single db sequence

**Comment by jahlfeld** [comment]
**What is wrong with the current solution?**

Let's say I have DB A and DB B. Both sequences start with value one.
Now user A saves an entity so the next sequence value is taken from DB A (now: SEQ_A = 2, SEQ_B = 1): commit_pk is set to Sequence.currentValue = 100. Now user B saves an entity (both requests run on the same application instance). But Sequence.currentValue is 101 with Sequence.currentLimit 199. So no sequence is touched an commit_pk for B's entity is set to 101. Depending on who does commit the commit that exceeds Sequence.currentLimit, one of both sequences is incremented, but this is pure randomness. If the next sequence value is taken from DB B, Sequence will again start with Sequence.currentValue 100 and Sequence.currentLimit 199, which means that the next 100 primary keys will have a risk of clashing with already existing primary keys.

Another problem:
When I have multiple instances of my application, the same problem can even occur with a single user / tenant, when different requests are answered by different application instances.

Summary:
Sequence is statefull (currentValue, currentLimit) and thus not suitable for distributed systems.

	Expected behavior:
	commit from A -> commit_pk is set to 100
	commit from B -> commit_pk is set to 100 too
	commit from A -> commit_pk is set to 101

	Actual behavior:
	commit from A -> commit_pk is set to 100
	commit from B -> commit_pk is set to 101
	commit from A -> commit_pk is set to 102

**How does my solution fix the problem?**

My solution solves this by providing a way to set a custom key generator (see PR). If needed, it is now possible to generate primary keys with an algorithm of your choice that does not use caching and works for distributed systems. The custom key generator is now used instead of the one that is provided by the dialect. If no key generator is provided in the SqlRepositoryBuilder, nothing will change.

**Comment by bartoszwalacik** [comment]
I still don't get it. A single javers instance writes to a single database. So you can have either 2 instances of javers connected to two different databases (so 2 separate environments) or 2 instances of javers connected to the same database.

**Comment by jahlfeld** [comment]
The latter one causes the problem I described under "another problem".
Let's say I have one database and one user, but two application instances. (replication -> reliablility, scaling, etc.)
Now my first request is answered by application instance A. When the load balancer directs my second request to application instance B, the JaVers instance on B will not know about the change on Sequence.currentValue of JaVers instance A, thus generating a pk that has already been used.

Btw, I don't think that using one JaVers instance for multiple databases should be a misuse.
First, there is the possibility to integrate the project via spring boot and normally, spring boot beans should be stateless.
Second, the choice of a database etc should be an implementation detail that should be provided by the connection manager but be irrelevant to other parts of the application.

But please read the first paragraph carefully - imho this is a serious problem unless having multiple replications of an application is a misuse, too.

**Comment by bartoszwalacik** [comment]
@jahlfeld please try to reduce your problem description to the simplest possible form (without introducing additional concepts, like request or load balancer, Javers has nothing to do with them).
So once again, are you trying to have a single javers instance writing to multiple databases? I mean one snapshot is written to DB A and a second to DB B? Obviously, javers won't be able to query from this kind of randomly distributed snapshot repository

**Comment by jahlfeld** [comment]
I'll try to reduce it to the simplest form, but I cannot do without the word "request" since this is closely connected with my problem.

I have ONE database and TWO application instances. Both instances are exactly the same and use one JaVers instance each. I have a client-server architecture. Both my applications need to be stateless as every request that comes from a client will be directed to one of those two application instances. I don't know which application instance will answer the request (and therefore save some entity and a commit with commit_pk) and honestly I don't care, because this is the point of stateless server applications.

I agree with you, that JaVers has nothing to do with requests or load balancing directly, but it prevents me from using multiple application instances:
If ONE user saves an entity, one application instance will create a JaVers commit and internally save a state by setting the fields Sequence.currentValue and Sequence.currentLimit.
My OTHER application instance will not know anything about these internal state changes.
If the SAME user saves another entity to the SAME DB and this is answered by the OTHER application instance (again: I don't care which application instance does the work), this instance will create a pk that has already been used by the first application instance.

Another solution to this problem would be to take EACH pk from the DB sequence instead of taking one sequence value and then calculating the next 100 pks without communicating with the DB.

**Comment by bartoszwalacik** [comment]
ok, so you have a normal, microservice-style deployment. Still, I don't see where is the problem. In SQL Repository , Commit_pk is generated using a DB sequence, which is safe to use in multi-instance, and multi-threaded apps. The trick with sequence allocation (delivering 100 PK values for a single seq value) looks safe for me.
Where exactly do you observe a bug? Can you reproduce it?

**Comment by ralphv13** [comment]
yes you are right. JaVers design is safe if you have one database structure to maintain. 
In our case we have a "multitenancy by schema design" and separate the auditing tables in each tenant schema. This makes it easier for us to discriminate the auditing entries for security reasons. Maybe this is beyond the initial design but the only issue we found so far is that the primary-key cache is initialized with the first sequence per tenant-based request and node/vm and that is not always the best choice in our scenario ;)
The solution @jahlfeld  try to implement is to decouple JaVers great auditing functionallity from the low level application design and to provide a extension point for this. 

If you need a reproduceable test-case for this issue, i try to create one in the upcoming week without using our own core-framework so it is more readable for you. just drop me a note.

thanks
Ralph

**Comment by bartoszwalacik** [comment]
@ralphv13 that's interesting, but how do you achieved "multitenancy by schema design" having a single Javers instance? A given Javers instance writes always to a single snapshot table.
